File: Leetcode/Typescript/2381. Shifting Letters II/solution.ts
// https://leetcode.com/problems/shifting-letters-ii/description

function shiftingLetters(s: string, shifts: number[][]): string {
  const diffArray = Array.from({ length: s.length }, (_) => 0);

  for (const [start, end, direction] of shifts) {
    if (direction) {
      diffArray[start] += 1;
      if (end + 1 < s.length) {
        diffArray[end + 1] -= 1;
      }
    } else {
      diffArray[start] -= 1;
      if (end + 1 < s.length) {
        diffArray[end + 1] += 1;
      }
    }
  }

  let moveCount = 0;
  const result: string[] = [];
  const A_CODE = "a".charCodeAt(0);

  for (let idx = 0; idx < s.length; idx++) {
    moveCount += diffArray[idx] % 26;
    if (moveCount < 0) {
      moveCount += 26;
    }

    result.push(
      String.fromCharCode(
        ((s.charCodeAt(idx) - A_CODE + moveCount) % 26) + A_CODE
      )
    );
  }

  return result.join("");
}

File: Leetcode/Python/2381. Shifting Letters II/solution.py
# https://leetcode.com/problems/shifting-letters-ii/description

class Solution:
    def shiftingLetters(self, s: str, shifts: List[List[int]]) -> str:
        length = len(s)
        diff_array = [0] * length

        for start, end, direction in shifts:
            if direction:
                diff_array[start] += 1

                if end + 1 < length:
                    diff_array[end + 1] -= 1
            else:
                diff_array[start] -= 1

                if end + 1 < length:
                    diff_array[end + 1] += 1

        move_count = 0
        result = []

        # In Python, the modulo operation automatically adjusts for negative numbers,
        # ensuring the result always falls within the range of the divisor, so there is no need to handle negatives explicitly.
        for idx in range(length):
            move_count += diff_array[idx]
            shifted_char = chr(
                (ord(s[idx]) - ord("a") + move_count) % 26 + ord("a"))
            result.append(shifted_char)

        return "".join(result)
